CMake - TEST_BIG_ENDIAN

**Describe the bug**

Adding 

```
include(TestBigEndian)
TEST_BIG_ENDIAN(IS_BIG_ENDIAN)
```

to the `CMakeLists.txt`fails. It reports:

> -- Check if the system is big endian
> -- Searching 16 bit integer
> -- Check size of unsigned short
> CMake Error at /usr/share/cmake-3.23/Modules/CheckTypeSize.cmake:146 (try_compile):
>   Cannot copy output executable
> 
>     ''
> 
>   to destination specified by COPY_FILE:
> 
>     'zephyr/coaps-zephyr/coaps-zephyr/build/CMakeFiles/CheckTypeSize/CMAKE_SIZEOF_UNSIGNED_SHORT.bin'
> 
>   Unable to find the executable at any of:
> 
>     zephyr/coaps-zephyr/coaps-zephyr/build/CMakeFiles/CMakeTmp/cmTC_13b71.elf
>     zephyr/coaps-zephyr/coaps-zephyr/build/CMakeFiles/CMakeTmp/Debug/cmTC_13b71.elf
>     zephyr/coaps-zephyr/coaps-zephyr/build/CMakeFiles/CMakeTmp/Development/cmTC_13b71.elf
> 
> Call Stack (most recent call first):
>   /usr/share/cmake-3.23/Modules/CheckTypeSize.cmake:277 (__check_type_size_impl)
>   /usr/share/cmake-3.23/Modules/TestBigEndian.cmake:55 (CHECK_TYPE_SIZE)
>   /usr/share/cmake-3.23/Modules/TestBigEndian.cmake:37 (__TEST_BIG_ENDIAN_LEGACY_IMPL)
>   zephyr/coaps-zephyr/modules/lib/tinydtls/zephyr/AutoConf.cmake:49 (TEST_BIG_ENDIAN)
>   zephyr/coaps-zephyr/modules/lib/tinydtls/zephyr/CMakeLists.txt:27 (include)
> 
> 
> -- Check size of unsigned int
> CMake Error at /usr/share/cmake-3.23/Modules/CheckTypeSize.cmake:146 (try_compile):
>   Cannot copy output executable
> 
>     ''
> 
>   to destination specified by COPY_FILE:
> 
>     'zephyr/coaps-zephyr/coaps-zephyr/build/CMakeFiles/CheckTypeSize/CMAKE_SIZEOF_UNSIGNED_INT.bin'
> 
>   Unable to find the executable at any of:
> 
>     zephyr/coaps-zephyr/coaps-zephyr/build/CMakeFiles/CMakeTmp/cmTC_6e8ef.elf
>     zephyr/coaps-zephyr/coaps-zephyr/build/CMakeFiles/CMakeTmp/Debug/cmTC_6e8ef.elf
>     zephyr/coaps-zephyr/coaps-zephyr/build/CMakeFiles/CMakeTmp/Development/cmTC_6e8ef.elf
> 
> Call Stack (most recent call first):
>   /usr/share/cmake-3.23/Modules/CheckTypeSize.cmake:277 (__check_type_size_impl)
>   /usr/share/cmake-3.23/Modules/TestBigEndian.cmake:60 (CHECK_TYPE_SIZE)
>   /usr/share/cmake-3.23/Modules/TestBigEndian.cmake:37 (__TEST_BIG_ENDIAN_LEGACY_IMPL)
>   zephyr/coaps-zephyr/modules/lib/tinydtls/zephyr/AutoConf.cmake:49 (TEST_BIG_ENDIAN)
>   zephyr/coaps-zephyr/modules/lib/tinydtls/zephyr/CMakeLists.txt:27 (include)
> 

**To Reproduce**

Add 

```
include(TestBigEndian)
TEST_BIG_ENDIAN(IS_BIG_ENDIAN)
```

to the `CMakeLists.txt`. Execute

>$ west build

**Expected behavior**

`TEST_BIG_ENDIAN(IS_BIG_ENDIAN)` sets `IS_BIG_ENDIAN` according the used processor.

**Impact**

`IS_BIG_ENDIAN` must be set manually,

**Environment (please complete the following information):**
 - Ubuntu 18.04
 - ncs 1.9.1

This seems to be more of a bug in the CMake itself than the Zephyr build system.

```
Determining if the system is big endian failed with the following output:
Change Dir: /home/stephanos/Dev/zephyrproject/zephyr/samples/hello_world/build/CMakeFiles/CMakeTmp

Run Build Command(s):/usr/bin/ninja cmTC_5e2a9 && [1/2] Building C object CMakeFiles/cmTC_5e2a9.dir/TestEndianess.c.obj
[2/2] Linking C executable cmTC_5e2a9
FAILED: cmTC_5e2a9
: && /opt/zephyr-sdk-0.14.1/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc   CMakeFiles/cmTC_5e2a9.dir/TestEndianess.c.obj -o cmTC_5e2a9   && :
/opt/zephyr-sdk-0.14.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/../../../../arm-zephyr-eabi/bin/ld: /opt/zephyr-sdk-0.14.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/../../../../arm-zephyr-eabi/lib/libc.a(lib_a-exit.o): in function `exit':
exit.c:(.text.exit+0x34): undefined reference to `_exit'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed
```

The compilation of `TestEndianness.c` fails because the toolchain, without any "required arguments," defaults to linking some libraries that expect certain symbols (e.g. `_exit` in case of `arm-zephyr-eabi`) to be defined by the application.

Note that the [CMake `TEST_BIG_ENDIAN` implementation calls `try_compile` without any arguments](https://github.com/Kitware/CMake/blob/8dbf38eaaf0d465a7b44e337204c3b763915ebb5/Modules/TestBigEndian.cmake#L91-L95), causing some required flags like `-nostdlib` to be not passed when compiling `TestEndianness.c`:

```cmake
     try_compile(HAVE_${VARIABLE}
      "${CMAKE_BINARY_DIR}"
      ${_test_file}
      OUTPUT_VARIABLE OUTPUT
      COPY_FILE "${CMAKE_BINARY_DIR}${CMAKE_FILES_DIRECTORY}/TestEndianess.bin" )
```

For instance, the compilation of `CMAKE_SIZEOF_UNSIGNED_SHORT.c`, also built as part of the `TEST_BIG_ENDIAN` function, succeeds because [the `CHECK_TYPE_SIZE` function calls `try_compile` with the `CMAKE_REQUIRED_FLAGS`](https://github.com/Kitware/CMake/blob/d985c12fa47b5bf4ff1f64be460d18af5e839e5c/Modules/CheckTypeSize.cmake#L150-L151):

```cmake
  try_compile(HAVE_${var} ${CMAKE_BINARY_DIR} ${src}
    COMPILE_DEFINITIONS ${CMAKE_REQUIRED_DEFINITIONS}
    LINK_OPTIONS ${CMAKE_REQUIRED_LINK_OPTIONS}
    LINK_LIBRARIES ${CMAKE_REQUIRED_LIBRARIES}
    CMAKE_FLAGS
      "-DCOMPILE_DEFINITIONS:STRING=${CMAKE_REQUIRED_FLAGS}"
...
```


Note that `-nostdlib` is added to `CMAKE_REQUIRED_FLAGS` here:

https://github.com/zephyrproject-rtos/zephyr/blob/04836af2a91f04d47431697ee1275ffb09e021df/cmake/compiler/gcc/target.cmake#L121

Applying the following patch to the CMake `TestBigEndian.cmake` makes this function run successfully:

```diff
--- /usr/share/cmake-3.22/Modules/TestBigEndian.cmake   2022-01-25 22:45:06.000000000 +0900
+++ cmaketest/Modules/TestBigEndian.cmake       2022-05-02 17:45:40.429774864 +0900
@@ -91,6 +91,9 @@
      try_compile(HAVE_${VARIABLE}
       "${CMAKE_BINARY_DIR}"
       ${_test_file}
+      CMAKE_FLAGS
+        "-DCOMPILE_DEFINITIONS:STRING=${CMAKE_REQUIRED_FLAGS}"
+        "-DINCLUDE_DIRECTORIES:STRING=${CMAKE_REQUIRED_INCLUDES}"
       OUTPUT_VARIABLE OUTPUT
       COPY_FILE "${CMAKE_BINARY_DIR}${CMAKE_FILES_DIRECTORY}/TestEndianess.bin" )
```

Output after applying the patch:
```
-- The C compiler identification is GNU 10.3.0
-- The CXX compiler identification is GNU 10.3.0
-- The ASM compiler identification is GNU
-- Found assembler: /opt/zephyr-sdk-0.14.1/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc
-- Check if the system is big endian
-- Searching 16 bit integer
-- Looking for sys/types.h
-- Looking for sys/types.h - found
-- Looking for stdint.h
-- Looking for stdint.h - found
-- Looking for stddef.h
-- Looking for stddef.h - found
-- Check size of unsigned short
-- Check size of unsigned short - done
-- Searching 16 bit integer - Using unsigned short
-- Check if the system is big endian - little endian
-- Configuring done
-- Generating done
```

Of course, the upstream CMake developers may argue that it is the toolchain's fault for not being able to produce an executable as-is without some special arguments (i.e. `-nostdlib`), which we can certainly address for the Zephyr SDK; but, other third-party GNU toolchains such as the GNU Arm Embedded also has this problem, and it is not without a good reason (i.e. the toolchain is configured such that it uses the newlib by default, which expects `_exit` to be defined either by the `libgloss` or the application).

Nevertheless, apart from the questionable toolchain issues, it is true that CMake has an inconsistency in its internal compiler checks that the `CMAKE_REQUIRED_INCLUDES` flags are used for some checks while they are not for the rest.

When @tejlmand (or @SebastianBoe) comes back, he can further comment on this and maybe file a bug report to the upstream CMake repository.

--- a/cmake/modules/FindTargetTools.cmake
+++ b/cmake/modules/FindTargetTools.cmake
@@ -60,6 +60,20 @@ set(CMAKE_SYSTEM_PROCESSOR ${ARCH})
 #   explicitly to specify the target system version.
 set(CMAKE_SYSTEM_VERSION ${PROJECT_VERSION})
 
+# https://cmake.org/cmake/help/latest/variable/CMAKE_LANG_BYTE_ORDER.html
+#   Byte order of <LANG> compiler target architecture, if known.
+#
+# Zephyr Kconfig defines BIG_ENDIAN according to arch, SoC, Board, so propagate
+# this setting to allow users to read the standard CMake variable or use
+# 'test_big_endian()' function.
+if(CONFIG_BIG_ENDIAN)
+  set(CMAKE_C_BYTE_ORDER   BIG_ENDIAN)
+  set(CMAKE_CXX_BYTE_ORDER BIG_ENDIAN)
+else()
+  set(CMAKE_C_BYTE_ORDER   LITTLE_ENDIAN)
+  set(CMAKE_CXX_BYTE_ORDER LITTLE_ENDIAN)
+endif()
+
 # We are not building dynamically loadable libraries
 set(BUILD_SHARED_LIBS OFF)
 

